File: createPersephoneBackupS3Script.py
#!/usr/bin/env python3
"""
createPersephoneBackupS3Script.py

This script performs a Restic backup by prompting for required configuration
values—including AWS credentials—and storing those values in a config file 
(.persephone.conf) for future runs. It manages four key values:
  - PERS_REPO_FILE: The file path to the repository file.
  - PERS_REPO_FILE_VALUE: The literal content (e.g. a repository URL/spec) that should be in that file.
  - PERS_PASSWD_FILE: The file path to the password file.
  - PERS_PASSWD_FILE_VALUE: The literal password that should be in that file.
Before running the backup, the script updates those files with the confirmed values.
Instead of directly running the backup, it outputs a minimal bash script in the format:

    #!/bin/bash
    export AWS_ACCESS_KEY_ID=<value>
    export AWS_SECRET_ACCESS_KEY=<value>
    restic -r <PERS_REPO_FILE_VALUE> --password-file <PERS_PASSWD_FILE> backup --verbose <backup_paths> --tag "<hostname>-$(date +\\%Y-\\%m-\\%d_\\%H-\\%M-\\%S)"
    echo ""
    echo "finis"

It then makes the script executable, creates /opt/persephone/ if needed, and moves the bash script there.
"""

import os
import subprocess
import getpass
import socket
import shutil

CONFIG_FILE = ".persephone.conf"
BASH_SCRIPT_NAME = "persephone.sh"
TARGET_DIR = "/opt/persephone/"

def load_config(config_file):
    """
    Loads configuration from the config file if it exists.
    Expected file format: one key="value" per line.
    Returns a dictionary with the configuration values.
    """
    config = {}
    if os.path.exists(config_file):
        with open(config_file, "r") as f:
            for line in f:
                line = line.strip()
                if not line or line.startswith("#"):
                    continue
                if "=" in line:
                    key, value = line.split("=", 1)
                    config[key.strip()] = value.strip().strip('"').strip("'")
    return config

def prompt_input(prompt_message, default_val=None, hidden=False):
    """
    Prompts the user for input with an optional default.
    If hidden=True, the input is hidden.
    """
    if default_val:
        prompt = f"{prompt_message} [{default_val}]: "
    else:
        prompt = f"{prompt_message}: "
    while True:
        if hidden:
            response = getpass.getpass(prompt).strip()
        else:
            response = input(prompt).strip()
        if response == "" and default_val is not None:
            return default_val
        elif response != "":
            return response
        else:
            print("Error: Input cannot be empty. Please enter a valid value.")

def get_confirmed_value(key, prompt_message, default_val, hidden=False):
    """
    Prompts for a value (e.g. a file path or literal value) and asks for confirmation.
    Returns the confirmed (or updated) value.
    """
    value = prompt_input(prompt_message, default_val, hidden)
    print(f"{key} is set to: {value}")
    answer = input("Is this correct? (Y/n): ").strip().lower()
    if answer in ["", "y", "yes"]:
        return value
    else:
        return prompt_input(f"Enter new value for {key}", default_val, hidden)

def save_config(config_file, config):
    """
    Saves the configuration dictionary to the config file.
    """
    with open(config_file, "w") as f:
        for key, value in config.items():
            f.write(f'{key}="{value}"\n')

def generate_bash_script(config):
    """
    Generates a minimal bash script (persephone.sh) using the configuration values.
    """
    hostname = socket.gethostname()
    # Build the restic command using the literal repository value and backup paths.
    bash_lines = [
        "#!/bin/bash",
        "",
        f"export AWS_ACCESS_KEY_ID={config['AWS_ACCESS_KEY_ID']}",
        f"export AWS_SECRET_ACCESS_KEY={config['AWS_SECRET_ACCESS_KEY']}",
        # Construct the restic backup command.
        f"restic -r {config['PERS_REPO_FILE_VALUE']} --password-file {config['PERS_PASSWD_FILE']} backup --verbose {config['BACKUP_PATHS_STR']} --tag \"{hostname}-$(date +\\%Y-\\%m-\\%d_\\%H-\\%M-\\%S)\"",
        "echo \"\"",
        "echo \"finis\""
    ]
    return "\n".join(bash_lines)

def main():
    # Load existing configuration if available.
    config = load_config(CONFIG_FILE)

    print("=== Restic Backup Configuration ===")

    # --- Repository file and its value ---
    default_repo_file = config.get("PERS_REPO_FILE", "/root/.persephone-repo")
    pers_repo_file = get_confirmed_value("PERS_REPO_FILE", "Enter the repository file path", default_repo_file)
    
    default_repo_value = config.get("PERS_REPO_FILE_VALUE", "s3:https://persephoneapi.domain.com/repo-name/endpoint-name")
    pers_repo_file_value = get_confirmed_value("PERS_REPO_FILE_VALUE", "Enter the repository file literal value", default_repo_value)

    # --- Password file and its value ---
    default_pass_file = config.get("PERS_PASSWD_FILE", "/root/.persephone-passwd")
    pers_pass_file = get_confirmed_value("PERS_PASSWD_FILE", "Enter the password file path", default_pass_file)
    
    default_pass_value = config.get("PERS_PASSWD_FILE_VALUE", "default-password")
    pers_pass_value = get_confirmed_value("PERS_PASSWD_FILE_VALUE", "Enter the password file literal value", default_pass_value, hidden=True)

    # --- Other configuration values ---
    default_backup_paths = config.get("BACKUP_PATHS_STR", "/root /home /var /etc /srv /usr /opt")
    backup_paths_str = prompt_input("Enter backup paths (space-separated)", default_backup_paths)

    default_aws_access_key = config.get("AWS_ACCESS_KEY_ID", "")
    aws_access_key = prompt_input("Enter AWS_ACCESS_KEY_ID", default_aws_access_key)

    default_aws_secret_key = config.get("AWS_SECRET_ACCESS_KEY", "")
    aws_secret_key = prompt_input("Enter AWS_SECRET_ACCESS_KEY", default_aws_secret_key, hidden=True)

    # Update configuration dictionary.
    config["PERS_REPO_FILE"] = pers_repo_file
    config["PERS_REPO_FILE_VALUE"] = pers_repo_file_value
    config["PERS_PASSWD_FILE"] = pers_pass_file
    config["PERS_PASSWD_FILE_VALUE"] = pers_pass_value
    config["BACKUP_PATHS_STR"] = backup_paths_str
    config["AWS_ACCESS_KEY_ID"] = aws_access_key
    config["AWS_SECRET_ACCESS_KEY"] = aws_secret_key

    # Save the configuration for future runs.
    save_config(CONFIG_FILE, config)
    print(f"\nConfiguration successfully saved to {CONFIG_FILE}")

    # Ensure the repository and password files on disk are updated with the literal values.
    try:
        with open(pers_repo_file, "w") as f:
            f.write(pers_repo_file_value + "\n")
        print(f"Updated repository file at {pers_repo_file} with the confirmed value.")
    except Exception as e:
        print(f"Error writing to {pers_repo_file}: {e}")
    
    try:
        with open(pers_pass_file, "w") as f:
            f.write(pers_pass_value + "\n")
        print(f"Updated password file at {pers_pass_file} with the confirmed value.")
    except Exception as e:
        print(f"Error writing to {pers_pass_file}: {e}")

    # Generate the minimal bash script.
    bash_script = generate_bash_script(config)
    script_path = os.path.join(os.getcwd(), BASH_SCRIPT_NAME)
    try:
        with open(script_path, "w") as f:
            f.write(bash_script + "\n")
        # Make it executable.
        os.chmod(script_path, 0o755)
        print(f"\nBash script '{BASH_SCRIPT_NAME}' generated in current directory.")
    except Exception as e:
        print(f"Error writing bash script: {e}")
        return

    # Create target directory if it doesn't exist and move the script there.
    try:
        os.makedirs(TARGET_DIR, exist_ok=True)
        target_path = os.path.join(TARGET_DIR, BASH_SCRIPT_NAME)
        shutil.move(script_path, target_path)
        print(f"Bash script moved to: {target_path}.") 
        print("")
        print("Please now run /opt/persephone/persephone.sh to check it works correctly.")
        print("")
        print("If it does work correctly, then please consider running ./createPersephoneSchedule.py to implement automated regular backups.")
    except Exception as e:
        print(f"Error moving bash script to {TARGET_DIR}: {e}")

if __name__ == "__main__":
    main()

File: unix/createPersephoneBackupS3Script.go
package main

import (
	"bufio"
	"fmt"
	"io"
	"io/ioutil"
	"log"
	"os"
	"path/filepath"
	"strings"
	"time"

	"golang.org/x/term"
)

const (
	CONFIG_FILE         = ".persephone.conf"
	BACKUP_SCRIPT_NAME  = "persephone.sh"
	INSPECT_SCRIPT_NAME = "persephoneInspectSnapshots.sh"
	TARGET_DIR          = "/opt/persephone/"
)

// loadConfig loads configuration from a file with one key="value" per line.
func loadConfig(configFile string) (map[string]string, error) {
	config := make(map[string]string)
	if _, err := os.Stat(configFile); os.IsNotExist(err) {
		return config, nil
	}

	data, err := ioutil.ReadFile(configFile)
	if err != nil {
		return nil, err
	}

	lines := strings.Split(string(data), "\n")
	for _, line := range lines {
		line = strings.TrimSpace(line)
		if line == "" || strings.HasPrefix(line, "#") {
			continue
		}
		if idx := strings.Index(line, "="); idx != -1 {
			key := strings.TrimSpace(line[:idx])
			val := strings.Trim(strings.TrimSpace(line[idx+1:]), `"'`)
			config[key] = val
		}
	}
	return config, nil
}

// promptInput prompts the user for input with an optional default value.
// If hidden is true, input is read without echoing.
func promptInput(promptMessage, defaultVal string, hidden bool) string {
	reader := bufio.NewReader(os.Stdin)
	var prompt string
	if defaultVal != "" {
		prompt = fmt.Sprintf("%s [%s]: ", promptMessage, defaultVal)
	} else {
		prompt = fmt.Sprintf("%s: ", promptMessage)
	}

	for {
		fmt.Print(prompt)
		var input string
		if hidden {
			byteInput, err := term.ReadPassword(int(os.Stdin.Fd()))
			fmt.Println()
			if err != nil {
				log.Printf("Error reading input: %v", err)
				continue
			}
			input = strings.TrimSpace(string(byteInput))
		} else {
			text, err := reader.ReadString('\n')
			if err != nil {
				log.Printf("Error reading input: %v", err)
				continue
			}
			input = strings.TrimSpace(text)
		}

		if input == "" && defaultVal != "" {
			return defaultVal
		} else if input != "" {
			return input
		} else {
			fmt.Println("Error: Input cannot be empty. Please enter a valid value.")
		}
	}
}

// saveConfig writes the configuration to the config file.
func saveConfig(configFile string, config map[string]string) error {
	var lines []string
	for key, value := range config {
		lines = append(lines, fmt.Sprintf(`%s="%s"`, key, value))
	}
	data := strings.Join(lines, "\n")
	return ioutil.WriteFile(configFile, []byte(data), 0644)
}

// generateBashScript generates the backup bash script content with --verbose=2 flag.
func generateBashScript(config map[string]string) string {
	hostname, _ := os.Hostname()
	lines := []string{
		"#!/bin/bash",
		"",
		fmt.Sprintf("export AWS_ACCESS_KEY_ID=%s", config["AWS_ACCESS_KEY_ID"]),
		fmt.Sprintf("export AWS_SECRET_ACCESS_KEY=%s", config["AWS_SECRET_ACCESS_KEY"]),
		fmt.Sprintf("restic -r %s --password-file %s backup --verbose=2 %s --tag \"%s-$(date +\\%%Y-\\%%m-\\%%d_\\%%H-\\%%M-\\%%S)\"",
			config["PERS_REPO_FILE_VALUE"], config["PERS_PASSWD_FILE"], config["BACKUP_PATHS_STR"], hostname),
		`echo ""`,
		`echo "finis"`,
	}
	return strings.Join(lines, "\n")
}

// generateInspectScript generates the inspect snapshots bash script content with --verbose=2 flag.
func generateInspectScript(config map[string]string) string {
	lines := []string{
		"#!/bin/bash",
		"",
		fmt.Sprintf("export AWS_ACCESS_KEY_ID=%s", config["AWS_ACCESS_KEY_ID"]),
		fmt.Sprintf("export AWS_SECRET_ACCESS_KEY=%s", config["AWS_SECRET_ACCESS_KEY"]),
		// Note: Added --verbose=2 flag after snapshots.
		fmt.Sprintf("restic -r %s --password-file %s snapshots --verbose=2", config["PERS_REPO_FILE_VALUE"], config["PERS_PASSWD_FILE"]),
		`echo ""`,
		`echo "Inspection complete."`,
	}
	return strings.Join(lines, "\n")
}

// backupFile creates a timestamped backup of the given file if it exists.
func backupFile(filePath string) error {
	if _, err := os.Stat(filePath); err == nil {
		timestamp := time.Now().Format("20060102-150405")
		backupPath := fmt.Sprintf("%s.backup.%s", filePath, timestamp)
		fmt.Printf("Backing up %s to %s...\n", filePath, backupPath)
		// Copy the file.
		source, err := os.Open(filePath)
		if err != nil {
			return err
		}
		defer source.Close()
		dest, err := os.Create(backupPath)
		if err != nil {
			return err
		}
		defer dest.Close()
		if _, err := io.Copy(dest, source); err != nil {
			return err
		}
		fmt.Println("Backup complete.")
	}
	return nil
}

func main() {
	// Load existing configuration if available.
	config, err := loadConfig(CONFIG_FILE)
	if err != nil {
		log.Fatalf("Error loading config: %v", err)
	}

	fmt.Println("=== Restic Backup Configuration ===")

	// --- Repository file and its literal value ---
	defaultRepoFile := "/root/.persephone-repo"
	if v, ok := config["PERS_REPO_FILE"]; ok && v != "" {
		defaultRepoFile = v
	}
	persRepoFile := promptInput("Enter the repository file path", defaultRepoFile, false)

	defaultRepoValue := "s3:https://persephoneapi.domain.com/repo-name/endpoint-name"
	if v, ok := config["PERS_REPO_FILE_VALUE"]; ok && v != "" {
		defaultRepoValue = v
	}
	persRepoFileValue := promptInput("Enter the repository file literal value", defaultRepoValue, false)

	// --- Password file and its literal value ---
	defaultPassFile := "/root/.persephone-passwd"
	if v, ok := config["PERS_PASSWD_FILE"]; ok && v != "" {
		defaultPassFile = v
	}
	persPassFile := promptInput("Enter the password file path", defaultPassFile, false)

	defaultPassValue := "default-password"
	if v, ok := config["PERS_PASSWD_FILE_VALUE"]; ok && v != "" {
		defaultPassValue = v
	}
	persPassValue := promptInput("Enter the password file literal value", defaultPassValue, true)

	// --- Other configuration values ---
	defaultBackupPaths := "/root /home /var /etc /srv /usr /opt"
	if v, ok := config["BACKUP_PATHS_STR"]; ok && v != "" {
		defaultBackupPaths = v
	}
	backupPathsStr := promptInput("Enter backup paths (space-separated)", defaultBackupPaths, false)

	defaultAWSAccessKey := ""
	if v, ok := config["AWS_ACCESS_KEY_ID"]; ok {
		defaultAWSAccessKey = v
	}
	awsAccessKey := promptInput("Enter AWS_ACCESS_KEY_ID", defaultAWSAccessKey, false)

	defaultAWSSecretKey := ""
	if v, ok := config["AWS_SECRET_ACCESS_KEY"]; ok {
		defaultAWSSecretKey = v
	}
	awsSecretKey := promptInput("Enter AWS_SECRET_ACCESS_KEY", defaultAWSSecretKey, true)

	// Update configuration map.
	config["PERS_REPO_FILE"] = persRepoFile
	config["PERS_REPO_FILE_VALUE"] = persRepoFileValue
	config["PERS_PASSWD_FILE"] = persPassFile
	config["PERS_PASSWD_FILE_VALUE"] = persPassValue
	config["BACKUP_PATHS_STR"] = backupPathsStr
	config["AWS_ACCESS_KEY_ID"] = awsAccessKey
	config["AWS_SECRET_ACCESS_KEY"] = awsSecretKey

	// Save the configuration for future runs.
	if err := saveConfig(CONFIG_FILE, config); err != nil {
		log.Fatalf("Error saving config: %v", err)
	}
	fmt.Printf("\nConfiguration successfully saved to %s\n", CONFIG_FILE)

	// Update repository file with the literal value.
	if err := ioutil.WriteFile(persRepoFile, []byte(persRepoFileValue+"\n"), 0644); err != nil {
		fmt.Printf("Error writing to %s: %v\n", persRepoFile, err)
	} else {
		fmt.Printf("Updated repository file at %s with the confirmed value.\n", persRepoFile)
	}

	// Update password file with the literal value.
	if err := ioutil.WriteFile(persPassFile, []byte(persPassValue+"\n"), 0644); err != nil {
		fmt.Printf("Error writing to %s: %v\n", persPassFile, err)
	} else {
		fmt.Printf("Updated password file at %s with the confirmed value.\n", persPassFile)
	}

	// Generate the backup bash script.
	backupScriptContent := generateBashScript(config)
	backupScriptPath := filepath.Join(".", BACKUP_SCRIPT_NAME)
	if err := ioutil.WriteFile(backupScriptPath, []byte(backupScriptContent+"\n"), 0755); err != nil {
		fmt.Printf("Error writing bash script: %v\n", err)
		return
	}
	fmt.Printf("\nBash script '%s' generated in current directory.\n", BACKUP_SCRIPT_NAME)

	// Generate the inspect snapshots bash script.
	inspectScriptContent := generateInspectScript(config)
	inspectScriptPath := filepath.Join(".", INSPECT_SCRIPT_NAME)
	if err := ioutil.WriteFile(inspectScriptPath, []byte(inspectScriptContent+"\n"), 0755); err != nil {
		fmt.Printf("Error writing inspect script: %v\n", err)
		return
	}
	fmt.Printf("Bash script '%s' generated in current directory.\n", INSPECT_SCRIPT_NAME)

	// Create target directory if it doesn't exist.
	if err := os.MkdirAll(TARGET_DIR, 0755); err != nil {
		fmt.Printf("Error creating target directory %s: %v\n", TARGET_DIR, err)
		return
	}

	// Move the scripts to the target directory.
	targetBackupPath := filepath.Join(TARGET_DIR, BACKUP_SCRIPT_NAME)
	targetInspectPath := filepath.Join(TARGET_DIR, INSPECT_SCRIPT_NAME)
	if err := os.Rename(backupScriptPath, targetBackupPath); err != nil {
		fmt.Printf("Error moving backup script: %v\n", err)
		return
	}
	if err := os.Rename(inspectScriptPath, targetInspectPath); err != nil {
		fmt.Printf("Error moving inspect script: %v\n", err)
		return
	}
	fmt.Printf("Bash scripts moved to: %s\n\n", TARGET_DIR)
	fmt.Printf("Please run %s to check the backup script works correctly.\n", targetBackupPath)
	fmt.Printf("And run %s to inspect your Persephone snapshots.\n\n", targetInspectPath)
	fmt.Println("If everything works correctly, consider running 'go run createPersephoneSchedule.go' to set up automated backups.")

	// Display final configuration for verification.
	fmt.Printf("\nFinal backup script (%s):\n", targetBackupPath)
	data, err := ioutil.ReadFile(targetBackupPath)
	if err != nil {
		fmt.Printf("Error reading %s: %v\n", targetBackupPath, err)
	} else {
		fmt.Println(string(data))
	}

	fmt.Printf("\nFinal inspect script (%s):\n", targetInspectPath)
	data, err = ioutil.ReadFile(targetInspectPath)
	if err != nil {
		fmt.Printf("Error reading %s: %v\n", targetInspectPath, err)
	} else {
		fmt.Println(string(data))
	}
}
